File: storage/native/storage/factory.c
#include "factory.h"
#include <memory/memory.h>

int32_t storage_factory_initialize(struct storage_factory* factory, struct storage_factory_configuration* configuration)
{
    float actual_alloc_factor;

    factory->memory = memory_create(configuration->quota_size, configuration->preallocation_size, configuration->slab_size);
    if (!factory->memory)
    {
        return -ENOMEM;
    }

    factory->storage_datas = calloc(1, sizeof(struct small_alloc));
    if (!factory->storage_datas)
    {
        return -ENOMEM;
    }
    small_alloc_create(&factory->storage_datas->allocator, &factory->memory->cache, 3 * sizeof(int), sizeof(uintptr_t), 1.05, &actual_alloc_factor);

    return 0;
}

const char* storage_create_string(struct storage_factory* factory, size_t size)
{
    return smalloc(&factory->storage_datas->allocator, size);
}

void storage_free_string(struct storage_factory* factory, const char* string, size_t size)
{
    smfree(&factory->storage_datas->allocator, (void*)string, size);
}

void storage_factory_destroy(struct storage_factory* factory)
{
    small_alloc_destroy(&factory->storage_datas->allocator);
    memory_destroy(factory->memory);
    free(factory->storage_datas);
    free(factory->memory);
}

File: storage/dart/lib/bindings/storage/factory.dart
// Generated
// ignore_for_file: unused_import

import 'dart:ffi';
import 'package:ffi/ffi.dart';
import '../../storage/bindings.dart';

final class storage_factory extends Struct {
  external Pointer<memory_instance> memory;
  external Pointer<memory_small_allocator> storage_datas;
}

final class storage_factory_configuration extends Struct {
  @Size()
  external int quota_size;
  @Size()
  external int slab_size;
  @Size()
  external int preallocation_size;
}

@Native<Int32 Function(Pointer<storage_factory> factory, Pointer<storage_factory_configuration> configuration)>(isLeaf: true)
external int storage_factory_initialize(Pointer<storage_factory> factory, Pointer<storage_factory_configuration> configuration);

@Native<Pointer<Utf8> Function(Pointer<storage_factory> factory, Size size)>(isLeaf: true)
external Pointer<Utf8> storage_create_string(Pointer<storage_factory> factory, int size);

@Native<Void Function(Pointer<storage_factory> factory, Pointer<Utf8> string, Size size)>(isLeaf: true)
external void storage_free_string(Pointer<storage_factory> factory, Pointer<Utf8> string, int size);

@Native<Void Function(Pointer<storage_factory> factory)>(isLeaf: true)
external void storage_factory_destroy(Pointer<storage_factory> factory);
